pages/XX-ai-methods/reproducible_prompts/Reproducible_Prompts.qmd

```diff
@@ -0,0 +1,299 @@
+    Additionally we have to specify the actual LLM because ollama can any model
+    you have installed.
+3.  `-m {}` allows you to provide additionall arguments to your LLM provider. In
+    our case we will be providing the default ollama port on local host
+    `"base_url":"http://localhost:11434"` so jupyter knows where to look for
+    ollama.
+
+```{python}
+%%ai ollama:llama3.2 -m {"base_url":"http://localhost:11434"}
+Introduce yourself
+```
+
+## Now let's try incoporating a python object
+We can incoporate a python object using the `{}` syntax inside our prompt. First
+Lets make a list in python.
+
+```{python}
+random_words = ["cat", "dog", "bird", "monkey", "elephant", "deer", "wolf", "bison", "elk"]  
+print(random_words)
+```
+
+Now we can embed this list in our prompt by calling the variable name inside of
+`{}`
+
+```{python}
+#| scrolled: true
+%%ai ollama:llama3.2 -m {"base_url":"http://localhost:11434"}
+Create a bullet list from the text of {random_words} in random order
+```
+
+We can set the output format using the `-f` flag. Lets specify we would like
+our output in json.
+
+```{python}
+%%ai ollama:llama3.2 -f json -m {"base_url":"http://localhost:11434"}
+Create a list from the text of {random_words} in random order
+```
+
+## Lets keep making things more reproducible
+We can Set a seed to try and get more reproducible results. This is an ollama
+parameter, so we provide it inside the `-m {}` arguments. Just add `"seed:42`
+after our `"base_url":"http://localhost:11434"` seperated by a comma. **Do not add any spaces**, as it can interfere with parsing the arguments.
+
+```{python}
+%%ai ollama:llama3.2 -m {"base_url":"http://localhost:11434","seed":42}
+Create a bullet list from the text of {random_words} in random order
+```
+
+## Let's keep tweaking the randomness
+- You can adjust to "Creativeness" of the responses with the `"temperature"` 
+- Default is 0.8, higher is more creative
+- You can set `"temperature":0` to give the least "creative" response which can help reproducibity.
+
+```{python}
+%%ai ollama:llama3.2 -m {"base_url":"http://localhost:11434","seed":42,"temperature":1.5}
+Create a list of 10 random words
+```
+
+```{python}
+%%ai ollama:llama3.2 -m {"base_url":"http://localhost:11434","seed":42,"temperature":0}
+Create a list of 10 random words
+```
+
+## What about passing the output to your python code?
+You may want to pass the output of a Juypter AI cell to your python code.
+The best way to do this is to call the contents of your output cell. First let's
+run an AI cell that outputs json. Python can then input the json as a python
+dictionary.
+
+```{python}
+%%ai ollama:llama3.2 -f json -m {"base_url":"http://localhost:11434","seed":42,"temperature":0}
+Create a list of 10 random words
+```
+
+To capture the output of this cell You can call the `Out[##]` function, substituting `##` for the cell number. This can be tricky because the cell
```

Comment:
Oh, I see this explanation now. I think you could probably skip using Out[] altogether -- it's a little strange to introduce it and then say we're actually going to use this other thing?

You can also use #| error: true at the top of the cell plus some code that doesn't work to get the page to display the error folks would see. That way it doesn't halt execution.

Reply:
Another idea would be to use a .column-margin element to put this bit off to the side. e.g.

::: {.column-margin title="Why not use `Out[]`?"}
Lorem ipsum...
:::